src/core/core.scale.js

```diff
@@ -90,10 +79,108 @@ function getPixelForGridLine(scale, index, offsetGridLines) {
 	return lineValue;
 }
 
-function computeTextSize(context, tick, font) {
-	return helpers.isArray(tick) ?
-		helpers.longestText(context, font, tick) :
-		context.measureText(tick).width;
+function measureText(ctx, data, gc, string) {
+	var key = ctx.font + ';' + string;
+	var textWidth = data[key];
+
+	if (!textWidth) {
+		textWidth = data[key] = ctx.measureText(string).width;
+		gc.push(key);
+	}
+	return textWidth;
+}
+
+function computeLabelSizes(ctx, tickFonts, ticks, cache) {
+	var length = ticks.length;
+	var widths = [];
+	var heights = [];
+	var data, gc, gcLen, i, j, jlen, tick, label, tickFont, width, height, nestedLabel, widest, highest;
+
+	cache = cache || {};
+	data = cache.data = cache.data || {};
+	gc = cache.garbageCollect = cache.garbageCollect || [];
+
+	for (i = 0; i < length; ++i) {
+		tick = ticks[i];
+		label = tick.label;
+		tickFont = tick.major ? tickFonts.major : tickFonts.minor;
+		width = height = 0;
+		ctx.font = tickFont.string;
+		// Undefined labels and arrays should not be measured
+		if (!helpers.isNullOrUndef(label) && !helpers.isArray(label)) {
+			width = measureText(ctx, data, gc, label);
+			height = tickFont.lineHeight;
+		} else if (helpers.isArray(label)) {
+			// if it is an array lets measure each element
+			for (j = 0, jlen = label.length; j < jlen; ++j) {
+				nestedLabel = label[j];
+				// Undefined labels and arrays should not be measured
+				if (!helpers.isNullOrUndef(nestedLabel) && !helpers.isArray(nestedLabel)) {
+					width = Math.max(width, measureText(ctx, data, gc, nestedLabel));
+					height += tickFont.lineHeight;
+				}
+			}
+		}
+		widths.push(width);
+		heights.push(height);
+	}
+
+	gcLen = gc.length / 2;
+	if (gcLen > length) {
+		for (i = 0; i < gcLen; ++i) {
+			delete data[gc[i]];
+		}
+		gc.splice(0, gcLen);
+	}
+
+	widest = widths.indexOf(Math.max.apply(null, widths));
+	highest = heights.indexOf(Math.max.apply(null, heights));
+
+	return {
+		first: {
+			width: widths[0] || 0,
+			height: heights[0] || 0
+		},
+		last: {
+			width: widths[length - 1] || 0,
+			height: heights[length - 1] || 0
+		},
+		widest: {
+			width: widths[widest] || 0,
+			height: heights[widest] || 0
+		},
+		highest: {
+			width: widths[highest] || 0,
+			height: heights[highest] || 0
+		}
+	};
+}
+
+function resolve(value1, value2, defaultValue) {
```

Comment:
I wonder if we could make `helpers.valueOrDefault` take an arbitrary number of parameters to avoid needing to add this `function`

Reply:
Should be easy to implement as we can iterate `arguments` but there is a performance penalty (if I recall correctly v1 and early v2 versions worked like that).

Perhaps we can special case the 2 version and then fallback to a loop when there are more than 2 args. Thoughts @simonbrunel?

Reply:
The code would be like this:
```js
valueOrDefault: function() {
	var i, ilen, value;
	for (i = 0, ilen = arguments.length - 1; i < ilen; ++i) {
		value = arguments[i];
		if (typeof value !== 'undefined') {
			return value;
		}
	}
	return arguments[i];
}
```
This can be used in other places like below replacing `custom.* ? custom.* : ...`.
https://github.com/chartjs/Chart.js/blob/b7139dbbef9ae978a5e26029bc1620b54c7652b4/src/controllers/controller.line.js#L72-L83
By the way, the code above has problems, for example, it falls back to `dataset.lineTension` when `custom.tension` is `false`. It can be fixed by using `valueOrDefault(custom.tension, dataset.lineTension, lineElementOptions.tension)`.

Reply:
I was planning to add a new core helper implementing @nagix suggestion but never had time to do it. I was thinking to call it `fallback(a, b, c ....)` which is shorter than `valueOrDefault` and I think more suitable.

@nagix you right, there is an issue with tension (and a few other, e.g. fill). All the part you highlighted will be replaced by scriptable options (thus, will rely on `helpers.options.resolve`), but until then, feel free to fix it. This new helper and these fixes could be also done in a new PR.

Reply:
@simonbrunel Do you think it would be better to use `fallback` for more than three arguments and keep using `valueOrDefault` for two arguments from performance perspective?

Reply:
> @simonbrunel Do you think it would be better to use `fallback` for more than three arguments and keep using `valueOrDefault` for two arguments from performance perspective?

I think yes (from performance perspective)
https://jsperf.com/fallback-vs-valueordefault/1

Reply:
You right, so we may want to keep using both: `valueOrDefault` for 2 values and `fallback` for more.

Reply:
#5965 is opened for this.